core/state_transition.go

```diff
@@ -242,6 +245,128 @@ func (st *StateTransition) to() common.Address {
+	if !ok {
+		panic("balanceOf method not found")
+	}
+
+	argument, err := method.Inputs.Pack(account)
+	if err != nil {
+		panic("failed to pack argument")
+	}
+
+	data := make([]byte, len(method.ID)+len(argument))
+	copy(data[0:len(method.ID)], method.ID)
+	copy(data[len(method.ID):], argument)
+	return data
+}
+
+func parseSoulBalanceResp(ret []byte) (*uint256.Int, error) {
+	method, ok := SoulGasTokenABI.Methods["balanceOf"]
+	if !ok {
+		panic("balanceOf method not found")
+	}
+
+	returnValueInterface, err := method.Outputs.Unpack(ret)
+	if err != nil {
+		return nil, fmt.Errorf("parseSoulBalanceResp Unpack failed:%w", err)
+	}
+
+	returnValue := new(big.Int)
+	err = method.Outputs.Copy(&returnValue, returnValueInterface)
+	if err != nil {
+		return nil, fmt.Errorf("parseSoulBalanceResp Copy failed:%w", err)
+	}
+	returnValueU256, overflow := uint256.FromBig(returnValue)
+	if overflow {
+		return nil, fmt.Errorf("parsed soul balance overflow:%v", returnValue)
+	}
+	return returnValueU256, nil
+}
+
+func burnSoulBalanceData(account common.Address, amount *big.Int) []byte {
+
+	method, ok := SoulGasTokenABI.Methods["burnFrom"]
+	if !ok {
+		panic("burnFrom method not found")
+	}
+
+	argument, err := method.Inputs.Pack(account, amount)
+	if err != nil {
+		panic("failed to pack argument")
+	}
+
+	data := make([]byte, len(method.ID)+len(argument))
+	copy(data[0:len(method.ID)], method.ID)
+	copy(data[len(method.ID):], argument)
+	return data
+}
+
+func mintSoulBalanceData(account common.Address, amount *big.Int) []byte {
+	method, ok := SoulGasTokenABI.Methods["batchMint"]
+	if !ok {
+		panic("batchMint method not found")
+	}
+
+	argument, err := method.Inputs.Pack([]common.Address{account}, []*big.Int{amount})
+	if err != nil {
+		panic("failed to pack argument")
+	}
+
+	data := make([]byte, len(method.ID)+len(argument))
+	copy(data[0:len(method.ID)], method.ID)
+	copy(data[len(method.ID):], argument)
+	return data
+}
+
+var (
+	callSoulGasLimit = uint64(8_000_000)
+	// use hardcoded DEPOSITOR_ACCOUNT both as minter and burner
+	depositorAddress = common.HexToAddress("0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001")
+)
+
+func (st *StateTransition) GetSoulBalance(account common.Address) (*uint256.Int, error) {
+	// this evm call is free of gas charging
+	ret, _, vmerr := st.evm.Call(vm.AccountRef(account), types.SoulGasTokenAddr, getSoulBalanceData(account), callSoulGasLimit, common.U2560)
+	if vmerr != nil {
+		return nil, vmerr
+	}
+	return parseSoulBalanceResp(ret)
+}
+
+func (st *StateTransition) SubSoulBalance(account common.Address, amount *big.Int) (err error) {
+	_, _, err = st.evm.Call(vm.AccountRef(depositorAddress), types.SoulGasTokenAddr, burnSoulBalanceData(account, amount), callSoulGasLimit, common.U2560)
+	if err == nil {
+		st.state.SubBalance(types.SoulGasTokenAddr, uint256.MustFromBig(amount))
+	}
+	return
+}
+
+func (st *StateTransition) AddSoulBalance(account common.Address, amount *big.Int) {
```

Comment:
Should we check whether the soul token is backed by gas token?

Reply:
Yeah the implementation is now assuming it's backed by gas token, I'll make it configurable in a follow-up commit.

Reply:
I will suggest at least leaving a TODO here.

Reply:
I just pushed a implementation that honors EnableSoulGasToken and IsSoulBackedByNative option from ChainConfig: https://github.com/ethstorage/op-geth/pull/1/commits/3985b7a44b3c86aeb047c6ee1fca78d41882c511